### ISSUE TITLE
Nodegit npm install error on FreeBSD 10

### ISSUE THREAD

**Issue by tankakatan** [nodegit/nodegit#690]
Hello!

How can i fix the following error?

<pre>
npm install nodegit
npm WARN package.json events@1.0.2 events is also the name of a node core module.
\
> nodegit@0.4.1 install /usr/local/www/president-kids.ru/node_modules/nodegit
> node lifecycleScripts/install

[nodegit] Fetching binary from S3.
[nodegit] Failed to install prebuilt binary, building manually.
[nodegit] Making sure dependencies are available and native code is generated
[nodegit] Something is missing, retrieving dependencies and regenerating code
combyne@0.8.1 node_modules/combyne

nan@1.9.0 node_modules/nan

tar@2.2.0 node_modules/tar
├── inherits@2.0.1
├── block-stream@0.0.8
└── fstream@1.0.7 (graceful-fs@3.0.8, mkdirp@0.5.1, rimraf@2.4.3)

mocha@2.2.5 node_modules/mocha
├── escape-string-regexp@1.0.2
├── supports-color@1.2.1
├── growl@1.8.1
├── commander@2.3.0
├── diff@1.4.0
├── debug@2.0.0 (ms@0.6.2)
├── mkdirp@0.5.0 (minimist@0.0.8)
├── glob@3.2.3 (inherits@2.0.1, graceful-fs@2.0.3, minimatch@0.2.14)
└── jade@0.26.3 (commander@0.6.1, mkdirp@0.3.0)

request@2.61.0 node_modules/request
├── aws-sign2@0.5.0
├── forever-agent@0.6.1
├── stringstream@0.0.4
├── caseless@0.11.0
├── oauth-sign@0.8.0
├── tunnel-agent@0.4.1
├── isstream@0.1.2
├── json-stringify-safe@5.0.1
├── extend@3.0.0
├── node-uuid@1.4.3
├── qs@4.0.0
├── combined-stream@1.0.5 (delayed-stream@1.0.0)
├── form-data@1.0.0-rc3 (async@1.4.2)
├── mime-types@2.1.5 (mime-db@1.17.0)
├── http-signature@0.11.0 (assert-plus@0.1.5, asn1@0.1.11, ctype@0.5.3)
├── bl@1.0.0 (readable-stream@2.0.2)
├── tough-cookie@2.0.0
├── hawk@3.1.0 (cryptiles@2.0.4, sntp@1.0.9, boom@2.8.0, hoek@2.14.0)
└── har-validator@1.8.0 (commander@2.8.1, chalk@1.1.1, is-my-json-valid@2.12.1, bluebird@2.9.34)

js-beautify@1.5.10 node_modules/js-beautify
├── nopt@3.0.3 (abbrev@1.0.7)
├── config-chain@1.1.9 (ini@1.3.4, proto-list@1.2.4)
└── mkdirp@0.5.1 (minimist@0.0.8)

lcov-result-merger@1.0.2 node_modules/lcov-result-merger
├── vinyl@0.4.6 (clone-stats@0.0.1, clone@0.2.0)
├── through2@0.6.5 (xtend@4.0.0, readable-stream@1.0.33)
└── vinyl-fs@0.3.13 (graceful-fs@3.0.8, defaults@1.0.2, strip-bom@1.0.0, mkdirp@0.5.1, glob-stream@3.1.18, glob-watcher@0.0.6)

coveralls@2.11.4 node_modules/coveralls
├── lcov-parse@0.0.6
├── log-driver@1.2.4
├── request@2.40.0 (aws-sign2@0.5.0, forever-agent@0.5.2, oauth-sign@0.3.0, stringstream@0.0.4, tunnel-agent@0.4.1, json-stringify-safe@5.0.1, qs@1.0.2, node-uuid@1.4.3, mime-types@1.0.2, form-data@0.1.4, http-signature@0.10.1, hawk@1.1.1, tough-cookie@2.0.0)
└── js-yaml@3.0.1 (argparse@0.1.16, esprima@1.0.4)

lodash@3.10.1 node_modules/lodash

nw-gyp@0.12.4 node_modules/nw-gyp
├── graceful-fs@2.0.3
├── osenv@0.1.3 (os-tmpdir@1.0.1, os-homedir@1.0.1)
├── nopt@2.2.1 (abbrev@1.0.7)
├── npmlog@0.1.1 (ansi@0.3.0)
├── semver@2.2.1
├── minimatch@0.4.0 (sigmund@1.0.1, lru-cache@2.6.5)
├── which@1.1.1 (is-absolute@0.1.7)
├── mkdirp@0.5.1 (minimist@0.0.8)
├── fstream@0.1.31 (inherits@2.0.1, graceful-fs@3.0.8)
├── tar@0.1.20 (inherits@2.0.1, block-stream@0.0.8)
├── glob@3.2.11 (inherits@2.0.1, minimatch@0.3.0)
└── rimraf@2.4.3 (glob@5.0.14)

istanbul@0.3.18 node_modules/istanbul
├── supports-color@1.3.1
├── which@1.0.9
├── abbrev@1.0.7
├── async@1.4.2
├── wordwrap@0.0.3
├── nopt@3.0.3
├── once@1.3.2 (wrappy@1.0.1)
├── mkdirp@0.5.1 (minimist@0.0.8)
├── esprima@2.5.0
├── resolve@1.1.6
├── fileset@0.2.1 (glob@5.0.14, minimatch@2.0.10)
├── escodegen@1.6.1 (esutils@1.1.6, estraverse@1.9.3, optionator@0.5.0, source-map@0.1.43, esprima@1.2.5)
├── js-yaml@3.4.0 (esprima@2.2.0, argparse@1.0.2)
└── handlebars@3.0.0 (optimist@0.6.1, source-map@0.1.43, uglify-js@2.3.6)

pangyp@2.3.0 node_modules/pangyp
├── which@1.0.9
├── rimraf@2.2.8
├── graceful-fs@3.0.8
├── osenv@0.1.3 (os-tmpdir@1.0.1, os-homedir@1.0.1)
├── nopt@3.0.3 (abbrev@1.0.7)
├── semver@4.3.6
├── mkdirp@0.5.1 (minimist@0.0.8)
├── fstream@1.0.7 (inherits@2.0.1)
├── minimatch@2.0.10 (brace-expansion@1.1.0)
├── tar@1.0.3 (inherits@2.0.1, block-stream@0.0.8)
├── glob@4.3.5 (inherits@2.0.1, once@1.3.2, inflight@1.0.4)
├── npmlog@1.0.0 (ansi@0.3.0, gauge@1.0.2, are-we-there-yet@1.0.4)
└── request@2.51.0 (caseless@0.8.0, forever-agent@0.5.2, aws-sign2@0.5.0, stringstream@0.0.4, oauth-sign@0.5.0, tunnel-agent@0.4.1, json-stringify-safe@5.0.1, node-uuid@1.4.3, qs@2.3.3, mime-types@1.0.2, combined-stream@0.0.7, bl@0.9.4, http-signature@0.10.1, form-data@0.2.0, hawk@1.1.1, tough-cookie@2.0.0)

jshint@2.8.0 node_modules/jshint
├── strip-json-comments@1.0.4
├── exit@0.1.2
├── shelljs@0.3.0
├── console-browserify@1.1.0 (date-now@0.1.4)
├── minimatch@2.0.10 (brace-expansion@1.1.0)
├── cli@0.6.6 (glob@3.2.11)
├── htmlparser2@3.8.3 (domelementtype@1.3.0, entities@1.0.0, domhandler@2.3.0, readable-stream@1.1.13, domutils@1.5.1)
└── lodash@3.7.0

[nodegit] Detecting vendor/libgit2/fb6df50b7f250a4fd8b2fab257f119a5185e9bf5.
[nodegit] Detecting vendor/libssh2/1.4.3.
[nodegit] Detecting vendor/http_parser/2.5.0.
[nodegit] Detecting generated code.
[nodegit] vendor/libgit2/fb6df50b7f250a4fd8b2fab257f119a5185e9bf5 already exists.
[nodegit] vendor/libssh2/1.4.3 already exists.
[nodegit] Generated code is intact.
[nodegit] Removing outdated vendor/http_parser/.
[nodegit] Retrieving vendor/http_parser/.
[nodegit] Successfully updated vendor/http_parser/2.5.0.
[nodegit] Everything is ready to go, attempting compilation
[nodegit] Building native node module.
{ [Error: Command failed: /bin/sh -c /usr/local/www/president-kids.ru/node_modules/nodegit/node_modules/.bin/pangyp rebuild
../vendor/libgit2/src/clone.c:479:20: warning: unused function 'repository_base' [-Wunused-function]
static const char *repository_base(git_repository *repo)
                   ^
1 warning generated.
../vendor/libssh2/src/session.c:52:10: fatal error: 'alloca.h' file not found
#include <alloca.h>
         ^
1 error generated.
gmake: *** [Release/obj.target/libssh2/vendor/libssh2/src/session.o] Error 1
gyp ERR! build error 
gyp ERR! stack Error: `gmake` failed with exit code: 2
gyp ERR! stack     at ChildProcess.onExit (/usr/local/www/president-kids.ru/node_modules/nodegit/node_modules/pangyp/lib/build.js:271:23)
gyp ERR! stack     at ChildProcess.emit (events.js:110:17)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (child_process.js:1074:12)
gyp ERR! System FreeBSD 10.0-RELEASE
gyp ERR! command "node" "/usr/local/www/president-kids.ru/node_modules/nodegit/node_modules/.bin/pangyp" "rebuild"
gyp ERR! cwd /usr/local/www/president-kids.ru/node_modules/nodegit
gyp ERR! node -v v0.12.2
gyp ERR! pangyp -v v2.3.0
gyp ERR! not ok 
]
  killed: false,
  code: 1,
  signal: null,
  cmd: '/bin/sh -c /usr/local/www/president-kids.ru/node_modules/nodegit/node_modules/.bin/pangyp rebuild' }
undefined
nodegit@0.4.1 node_modules/nodegit
├── promisify-node@0.1.5 (nodegit-promise@1.0.2)
├── nodegit-promise@2.0.1 (asap@2.0.3)
├── which-native-nodish@1.1.3 (find-parent-dir@0.3.0, nodegit-promise@1.0.2)
├── fs-extra@0.18.4 (jsonfile@2.2.1, graceful-fs@3.0.8, rimraf@2.4.3)
└── npm@2.14.0
</pre>

**Comment by tankakatan** [comment]
i've already tried to reinstall: node, npm, pangyp, nw-gyp, node-gyp – nothing helped

**Comment by tgallant** [comment]
@tankakatan 

I ran into the same error on FreeBSD 10 and found a solution. Its not a great solution at the moment as my knowledge of Makefiles and C preprocessing isn't great, but hopefully someone can use this info to create a workable solution.

### My Solution

1. Go into `node_modules/nodegit`
2. Open the file `vendor/libssh2/src/session.c` and delete line 52.
3. Set the environment variable `CXX` to `g++49`  (or whatever your c++ compiler is). I'm using tcsh so it looked like this `setenv CXX g++49`. For Bash it would be `export CXX=g++49`.
4. Run `npm install`

everything should compile properly at this point and you should be able to run your application.

### More Info

1. `alloca.h` is not used in freebsd. `alloca()` is defined in `stdlib.h` on FreeBSD.

The error is thrown at `vendor/libssh2/src/session.c:52`. the ifdef is checking the value of `HAVE_ALLOCA_H`. This needs to be false on FreeBSD. Not sure how or where to set that.

2. Users will have to have the `CXX` environment variable set or the Makefile will have to have some way of finding the right compiler.

Let me know if this method works for you or if you have any questions

**Comment by saper** [comment]
Interesting, I have tried to compile nodegit with clang on FreeBSD 10.1 with #702 and #708 included and it compiled fine.

**Comment by maxkorp** [comment]
Have you guys checked that without the change to session.c? Just bumping the c++ compiler?

**Comment by saper** [comment]
> Have you guys checked that without the change to session.c? Just bumping the c++ compiler?

Not really - I think I am using similar environment. Couldn't reproduce the original
problem.

**Comment by maxkorp** [comment]
Yeah I'd be curious to see if @tgallant gets functional results with just the G++ change.

**Comment by saper** [comment]
I actually wonder since 10.x ships with clang by default. Maybe @tgallant should try `env CC=cc CXX=c++ npm install`. 

I am using

````
FreeBSD clang version 3.4.1 (tags/RELEASE_34/dot1-final 208032) 20140512
Target: x86_64-unknown-freebsd10.1
Thread model: posix
````

**Comment by maxkorp** [comment]
I mean, IIRC gcc/g++ are just links to clang on OSX, and it builds fine for me on OSX, but it's hard to say.

**Comment by saper** [comment]
> I mean, IIRC gcc/g++ are just links to clang on OSX, and it builds fine for me on OSX, but it's hard to say. 
They mention explicitly setting CXX to gcc49 which is gcc 4.9 installed from ports.

**Comment by maxkorp** [comment]
Ya, i was more referring to your thought that the default compiler is clang which would theoretically be causing the issue, but pointing out on OSX clang works. Who knows how the versions line up however. Like, if changing to gcc4.9 works, which version of clang are they using that doesnt, because i compile successfully with clang on OSX.

**Comment by saper** [comment]
> Ya, i was more referring to your thought that the default compiler is
> clang which would theoretically be causing the issue, but pointing out
> on OSX clang works. Who knows how the versions line up however. Like,
> if changing to gcc4.9 works, which version of clang are they using
> that doesnt, because i compile successfully with clang on OSX.

I can now confirm I have compiled nodegit successfully with

    env CC=gcc49 CXX=g++49 npm install

I only had to add /usr/local/lib/gcc49 to LD_LIBRARY_PATH
to make sure gcc49's libstdc++.so.6 is available.

**Comment by maxkorp** [comment]
Yeah, but you weren't you already able to compile correctly? I'd be more curious to see @tankakatan's and @tgallant's results with just the compiler change (without changing the libgit2 source) since they were otherwise unable to compile.

**Comment by saper** [comment]
Well, I tried hard enough to reproduce it. clang and g++ use different sets of C++ libraries and share only some part of headers; I think it is now submitters turn. Maybe they didn's set `CC` to `gcc49`, I don't know.

My system does not have `<alloca.h>` at all.

Before we continue I think we should ask OP to present their `vendor/libssh2/config.log`, we have not enough facts.

**Comment by tgallant** [comment]
hey all, sorry for the late reply, i updated my nodegit dependency to 0.5.0 and can verify that it compiles both using clang and gcc on FreeBSD 10.2-STABLE.

my old package.json was pulling in version 0.4.1 and i was still getting the same alloca.h error when i tried compiling with clang.

so for me at least, everything just works on 0.5.0

**Comment by saper** [comment]
> hey all, sorry for the late reply, i updated my nodegit dependency to 0.5.0 and can verify that it compiles both using clang and gcc on FreeBSD 10.2-STABLE.
> 
> my old package.json was pulling in version 0.4.1 and i was still getting the same alloca.h error when i tried compiling with clang.
> 
> so for me at least, everything just works on 0.5.0

Great! good to hear that!
Have a nice flight with node-git on FreeBSD!

**Comment by johnhaley81** [comment]
Closing this out as it seems fixed in 0.5.0
